### PR TITLE
Make sure invalid RegExp string is properly sanitized in query

### LINKED ISSUE #1781: Unsanitized RegExp special content crashes application server
## Steps to reproduce

Just issue a query using any of the availalbe regexp operators: **like, nlike, ilike, nilike, regexp** .
For example lets say we have a model **Person** as follows:

```json
name: 'string',
lastName: 'string',
description: 'string'
```
If we issue a query of the form `Person.find({ where: { name: '[' } });` the application will crash

## Current Behavior

Issuing an improper query (also form remote API request) will result i an application crash with the Error **SyntaxError: Invalid regular expression: /[/: Unterminated character class`**

## Expected Behavior

The query should be sanitized first in order to be able to use special characters such as `[`

## Additional information

linux x64 10.15.3
loopback-datasource-juggler@4.13.0 /home/mitsos/Documents/Projects/loopback-datasource-juggler
├── loopback-connector@4.9.0

## Related Issues

Same issue here: https://github.com/strongloop/loopback-datasource-juggler/issues/1570
It was closed as  stale, but bug still persists

_See [Reporting Issues](http://loopback.io/doc/en/contrib/Reporting-issues.html) for more tips on writing good issues_

### PR DESCRIPTION
**Pull request by mitsos1os**
Fix the case where an invalid RegExp string is queried on a model. Instead of crashing the app,
we first check if the query can be successfully converted to RegExp and if not, it is sanitized.

Fixes #1781 

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-datasource-juggler) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](https://loopback.io/doc/en/contrib/style-guide-es6.html)
- [x] Commit messages are following our [guidelines](https://loopback.io/doc/en/contrib/git-commit-messages.html)

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -317,6 +318,32 @@ function isProhibited(key, prohibitedKeys) {
   return false;
 }
 
+/**
+ * Accept an operator key and return whether it is used for a regular expression query or not
+ * @param {string} operator
+ * @returns {boolean}
+ */
+function isRegExpOperator(operator) {
```
Comment (emonddr) [inline lib/utils.js]: what about other operators?

https://loopback.io/doc/en/lb4/apidocs.repository.operators.html

image

File: test/basic-querying.test.js [hunk 2]
```diff
@@ -591,6 +592,12 @@ describe('basic-querying', function() {
             });
         });
 
+      it('should sanitize invalid usage of like', async () => {
```
Comment (emonddr) [inline test/basic-querying.test.js]: Why should an invalid usage of find with '[' return an actual result?

Comment (emonddr) [reply test/basic-querying.test.js]: I am confused because another test below returns `empty`. This is what I would expect here as well.

e.g.

image

**Comment by mitsos1os** [comment]
Hi @bajtos and thank you for your quick reactions to the pull request. 

I understand your expressed concern about functionality.

To be sincere, this has come to our attention because we are using a web app which in a user view provides some DB lists.
In that view, there is a field for *easy-search* that accepts user input and searches all available list fields in the database using **like** operator to be able to partial match field contents.
This is where we caught the bug. There was a list element with field content starting with **[type]**. So when we tried inserting `[` in the **easy-search** input, the query sent was `like: '['` which caused the aforementioned application crash in the issue report. 

In our case, it is a matter of user-spec if we want to enable RegExp special characters from **easy-search**, or provide an easier way of direct partial-searching the user input by escaping all user input. I leave this up to you to decide what you want me to implement.

What **needs** to be done, is for the application not to crash.
So we could just sent a desired message and let the actual handling in the user application

File: lib/utils.js (hunk 1 above)
Comment (raymondfeng) [reply lib/utils.js]: @emonddr This is for regexp related operators only.

Comment (mitsos1os) [reply lib/utils.js]: @emonddr as @raymondfeng said this targets the crash caused by Invalid regular expression operators

File: test/basic-querying.test.js (hunk 2 above)
Comment (mitsos1os) [reply test/basic-querying.test.js]: These tests were made under the context described in my reponse to @bajtos comment.  I approached it by choosing to interpret an invalid RegExp string as a literal (escaped) one, which seraches for the actual characters provided.

The first test returns the result because of the regex search on new **tag** field which matches the `[singer]` content, while the `ilike` is on the name field which does not match.

If you guys decide that you just simply want to return the error as discussed with @bajtos then I will have to change these tests.

**Comment by mitsos1os** [comment]
any news with the pull request? Should I proceed with the forementioned functionality?

File: lib/utils.js [hunk 3]
```diff
@@ -317,6 +318,32 @@ function isProhibited(key, prohibitedKeys) {
...
+ * @param {string} str
+ * @returns {string}
+ */
+function escapeRegExp(str) {
+  assert.strictEqual(typeof str, 'string', 'String required for regexp escaping');
+  let success = true;
+  try {
+    new RegExp(str); // try to parse string as regexp
+  } catch (unused) {
+    success = false;
```
Comment (bajtos) [inline lib/utils.js]: While I am not very happy about the current implementation, I also don't want to block fixing the source of crashes. 

To make it easier for us to change this behavior in the future and start rejecting requests with unescaped regexp values, I am proposing to print a warning here. For example:

```js
  try {
     new RegExp(str); // try to parse string as regexp
   } catch (unused) {
     console.warn(
       'Auto-escaping invalid RegExp value %j supplied by the caller. ' + 
         'Please note this behavior may change in the future.', 
       str
     );
     success = false;
    }
```

While making changes in this function, can we simplify the implementation as follows?

```js
function escapeRegExp(str) {
  assert.strictEqual(typeof str, 'string', 'String required for regexp escaping');
  try {
    new RegExp(str); // try to parse string as regexp
    return str;
  } catch (unused) {
    console.warn(/*...*/);
    return str.replace(/[\-\[\]\/\{\}\(\)\+\?\.\\\^\$\|]/g, '\\$&');
  }
}
```

File: test/basic-querying.test.js [hunk 4]
```diff
@@ -616,6 +623,11 @@ describe('basic-querying', function() {
           done();
         });
       });
+
+      it('should properly sanitize invalid ilike filter', async () => {
+        const users = await User.find({where: {name: {ilike: '['}}});
+        users.should.be.empty;
```
Comment (bajtos) [inline test/basic-querying.test.js]:
```suggestion
        users.should.be.empty();
```

File: lib/utils.js (hunk 3 above)
Comment (mitsos1os) [reply lib/utils.js]: I can change it right now, because we also want this fix to land.
Would you prefer, just catching the error, and returning it as a SyntaxError: Invalid regular expression to the request?

Comment (mitsos1os) [reply lib/utils.js]: well? 😃

**Comment by dhmlau** [comment]
@slnode test please

File: lib/utils.js (hunk 3 above)
Comment (jannyHou) [reply lib/utils.js]: I was following the chat here and +1 for @bajtos 's simplified implementation:

```js
function escapeRegExp(str) {
  assert.strictEqual(typeof str, 'string', 'String required for regexp escaping');
  try {
    new RegExp(str); // try to parse string as regexp
    return str;
  } catch (unused) {
    // we can reject with proper message in a follow-up PR
    console.warn(/*...*/);
    return str.replace(/[\-\[\]\/\{\}\(\)\+\?\.\\\^\$\|]/g, '\\$&');
  }
}
```

File: test/basic-querying.test.js [hunk 5]
```diff
@@ -23,6 +23,7 @@ describe('basic-querying', function() {
       birthday: {type: Date, index: true},
       role: {type: String, index: true},
       order: {type: Number, index: true, sort: true},
+      notes: {type: String, index: true},
```
Comment (jannyHou) [inline test/basic-querying.test.js]: @mitsos1os Wouldn't the property here be `tag` instead of `notes`? The connector's tests fail due to `tag` is not defined in the model(therefore no mapping column in the mysql/mssql schema).

File: test/memory.test.js [hunk 6]
```diff
@@ -201,6 +201,7 @@ describe('Memory connector', function() {
       birthday: {type: Date, index: true},
       role: {type: String, index: true},
       order: {type: Number, index: true, sort: true},
+      notes: {type: String, index: true},
```
Comment (jannyHou) [inline test/memory.test.js]: ditto, I think the property should be `tag` not `notes`

File: lib/utils.js (hunk 3 above)
Comment (mitsos1os) [reply lib/utils.js]: ok @jannyHou! changing it now!

File: test/memory.test.js (hunk 6 above)
Comment (mitsos1os) [reply test/memory.test.js]: Yes you are right! fixed!

File: test/basic-querying.test.js (hunk 5 above)
Comment (mitsos1os) [reply test/basic-querying.test.js]: Fixed!

**Comment by mitsos1os** [comment]
@jannyHou @bajtos All changes made! Please check again so we can proceed

**Comment by dhmlau** [comment]
@mitsos1os, could you please squash your commits? Thanks! 

@jannyHou @emonddr, could you please review again?

**Comment by mitsos1os** [comment]
> @mitsos1os, could you please squash your commits? Thanks!
> 
> @jannyHou @emonddr, could you please review again?

@dhmlau done as well!

**Comment by mitsos1os** [comment]
> @mitsos1os, could you please squash your commits? Thanks!
> 
> @jannyHou @emonddr, could you please review again?

@dhmlau done as well!

**Comment by jannyHou** [comment]
@slnode test please

**Comment by dhmlau** [comment]
@mitsos1os, thanks for your contribution.  Your PR has landed! 🎉

**Comment by dhmlau** [comment]
`loopback-datasource-juggler@4.14.0` has been released.

**Comment by mitsos1os** [comment]
> `loopback-datasource-juggler@4.14.0` has been released.

Will this be backported to 3.x as well? We are using it in Loopback v3 currently

**Comment by dhmlau** [comment]
Could you please submit a patch against the 3.x branch?

**Comment by mitsos1os** [comment]
@dhmlau There you go:
https://github.com/strongloop/loopback-datasource-juggler/pull/1785
